### PR TITLE
feat(ui): update Portal

### PR DESCRIPTION
**Pull request by franzheidl**
# feat(ui): update Portal

Make our portals more robust and isolate them against each other. Avoid conflicts and make positioning portaled content more robust. 
We need this as a basis to migrate more components to use the portal (`ContextMenu`, `DateTimePicker`), so we can then address positioning issues we currently have on a common basis.

This PR introduces individual parent divs per each individual portal in a parent root element that is rendered by `PortalProvider`:

```
<div class="juno-portal-root" id="juno-portal-root">
    <div class="juno-portal">
        <!-- portal is mounted here -->
    </div>
        <div class="juno-portal">
        <!-- another portal is mounted here -->
    </div>
</div>
```

Users can use Portals in the same way as before, i.e. either as a `<PortalProvider.Portal />` component, or by calling the `usePortalRef` hook and create a portal in their components manually that mounts at the node returned by the hook. 

```
// Component
<PortalProvider>
    <PortalProvider.Portal>
        <MyComponent />
    <PortalProvider.Portal>
</PortalProvider>

// Hook
const portalNode = usePortalRef

createPortal(<MyComponent />, portalNode)
```

AppShell makes sure PortalProvider is rendered in the correct place of the component tree so that all elements rendered into portals have access to the Juno styles.

`PortalProvider` also exports a `PortalContext` that contains a ref to the portal root node as its sole value.

# Changes Made

- Update `PortalProvider` to render a portal root element in our apps. 
- Update `PortalProvider.Portal` component and `usePortalRef hook` to wrap portaled content into individual divs

# Related Issues

- Issue 1: [link to issue]

# Screenshots (if applicable)

# Testing Instructions

1. `npm i`
2. `npm run TASK`

# Checklist

- [ ] I have performed a self-review of my code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have made corresponding changes to the documentation (if applicable).
- [ ] My changes generate no new warnings or errors.

### CODE REVIEW

**Review by barsukov — COMMENTED** [review]
This component is kinda blocking us for typescript migration cause staff that used there, also used in many other spots. If you do not mind I would easily transfer it to typescript as well.

File: packages/ui-components/src/components/PortalProvider/PortalProvider.component.js [hunk 1]
```diff
@@ -1,79 +1,142 @@
...
-  return ref ? createPortal(children, ref) : null
-}
+    return () => {
+      // Clean up the portal element when unmounting:
+      if (containerRef.current && rootRef?.current) {
+        rootRef.current.removeChild(containerRef.current)
+        containerRef.current = null
+      }
+    }
+  }, [rootRef])
```
Comment (barsukov) [inline packages/ui-components/src/components/PortalProvider/PortalProvider.component.js:102]: by ading so many logic inside the useEffect you violate this hook rule 
https://github.com/facebook/react/blob/main/packages/eslint-plugin-react-hooks/README.md
Which called **react-hooks/exhaustive-deps**
For more info please check
https://stackoverflow.com/questions/58866796/understanding-the-react-hooks-exhaustive-deps-lint-rule

That has been not enabled by this project yet, but if we do we need to get reed of the code that you provided.

File: packages/ui-components/src/components/PortalProvider/PortalProvider.component.js [hunk 2]
```diff
@@ -1,79 +1,142 @@
...
     </PortalContext.Provider>
   )
 }
-// bind Portal to PortalProvider
+
+// Bind Portal to PortalProvider:
 PortalProvider.Portal = Portal
 Portal.displayName = "PortalProvider.Portal"
 
 PortalProvider.propTypes = {
```
Comment (barsukov) [inline packages/ui-components/src/components/PortalProvider/PortalProvider.component.js:130]: My suggestion of typescript that propTypes are going to be deprecated by React 19 we have to move on. That's why we need this component on Typescript

File: packages/ui-components/src/components/PortalProvider/PortalProvider.test.js [hunk 3]
```diff
@@ -1,40 +1,78 @@
...
     )
-    expect(document.querySelector(".juno-portal-container")).toBeInTheDocument()
+    expect(document.getElementById("juno-portal-root")).toBeInTheDocument()
+    expect(screen.getByRole("button")).toBeInTheDocument()
+    const parentEl = screen.getByRole("button").parentElement
+    expect(parentEl).toHaveClass("juno-portal")
+    expect(parentEl.parentElement).toHaveClass("juno-portal-root")
+  })
+
+  test("renders a portal root container with an id as passed", async () => {
```
Comment (barsukov) [inline packages/ui-components/src/components/PortalProvider/PortalProvider.test.js:35]: no async without await please

File: packages/ui-components/src/components/PortalProvider/PortalProvider.test.js [hunk 4]
```diff
@@ -1,40 +1,78 @@
...
 describe("PortalProvider", () => {
-  test("renders a PortalProvider wrapper div", async () => {
+  test("renders a portal root container", async () => {
+    render(<PortalProvider />)
+    const portalRoot = document.getElementById("juno-portal-root")
+    expect(portalRoot).toBeInTheDocument()
+    expect(portalRoot).toHaveClass("juno-portal-root")
+  })
+
+  test("renders children wrapped into a .juno-portal container into a portal using PortalProvider.Portal", async () => {
```
Comment (barsukov) [inline packages/ui-components/src/components/PortalProvider/PortalProvider.test.js:20]: same here

File: packages/ui-components/src/components/PortalProvider/PortalProvider.test.js [hunk 5]
```diff
@@ -1,40 +1,78 @@
...
       </PortalProvider>
     )
-    expect(document.querySelector(".juno-portal-container")).toBeInTheDocument()
-    expect(screen.getByRole("button")).toBeInTheDocument()
-    expect(screen.getByRole("button")).toHaveTextContent("Hello")
+    const portalContainers = document.querySelectorAll(".juno-portal")
+    expect(portalContainers.length).toBe(3)
+  })
+
+  test("renders an additional custom className as passed", async () => {
```
Comment (barsukov) [inline packages/ui-components/src/components/PortalProvider/PortalProvider.test.js:71]: same

File: packages/ui-components/src/components/PortalProvider/index.js [hunk 6]
```diff
@@ -1,6 +1 @@
-/*
```
Comment (barsukov) [inline packages/ui-components/src/components/PortalProvider/index.js:1]: please return the licenses I guess it is needed

File: packages/ui-components/src/components/PortalProvider/PortalProvider.test.js [hunk 7]
```diff
@@ -1,40 +1,78 @@
...
-    expect(document.querySelector("#my-portal-provider")).toBeInTheDocument()
+    expect(document.getElementById("juno-portal-root")).toBeInTheDocument()
+    expect(screen.getByRole("button")).toBeInTheDocument()
+    const parentEl = screen.getByRole("button").parentElement
+    expect(parentEl).toHaveClass("juno-portal")
+    expect(parentEl.parentElement).toHaveClass("juno-portal-root")
   })
 
-  test("renders other children as passed", async () => {
+  test("renders individual juno portal containers for multiple portals", async () => {
```
Comment (barsukov) [inline packages/ui-components/src/components/PortalProvider/PortalProvider.test.js:53]: same

**Review by barsukov — CHANGES_REQUESTED** [review]
We have to move on and make this as a PR, would be nice to transfer it to TS as soon as possible if you need any support please let me know.

File: packages/ui-components/src/components/PortalProvider/index.js (hunk 6 above)
Comment (franzheidl) [reply packages/ui-components/src/components/PortalProvider/index.js]: Done.

**Review by andypf — APPROVED** [review]
lgtm

**Pull request merged**